gr-uhd/lib/usrp_sink_impl.cc

```diff
@@ -187,6 +187,46 @@ ::uhd::gain_range_t usrp_sink_impl::get_gain_range(const std::string& name, size
     return _dev->get_tx_gain_range(name, chan);
 }
 
+bool usrp_sink_impl::has_power_reference(const size_t chan)
+{
+#ifdef UHD_USRP_MULTI_USRP_POWER_LEVEL
+    const size_t dev_chan = _stream_args.channels[chan];
+    return _dev->has_tx_power_reference(dev_chan);
+#else
+    return false;
+#endif
+}
+
+void usrp_sink_impl::set_power_reference(const double power_dbm, const size_t chan)
+{
+#ifdef UHD_USRP_MULTI_USRP_POWER_LEVEL
+    const size_t dev_chan = _stream_args.channels[chan];
+    _dev->set_tx_power_reference(power_dbm, dev_chan);
+#else
+    throw std::runtime_error("not implemented in this version");
```

Comment:
After some thought -- me trying to decide whether the `throw` is the way to go here, or a warning & still let the user's code execute -- I have settled on the former (the `throw` as you use) ... but with a more verbose message, which can be printed out before the `throw` or as part of the `throw` as you see fit: something to the effect of "The code being executed is trying to use the UHD power reference functionality, which was introduced in UHD 4.0. The actual UHD version being used is 3.15 or earlier, and does not provide this functionality. Please either update the installed UHD to version 4.0 or later, or find a way for the code being executed to conditionally use the power reference."

Reply:
@michaelld I tried to make these functions look like the rest of this file. There's a lot of other APIs that have this exact format, which I followed for these.

That's not to say I don't agree with your review comments, but I have follow-up questions:
- Are you comments taking the rest of the file into consideration? E.g., LO controls
- Your message is longer than typical exception messages. Should that go to the logger, or into the exception?
- On the "still let the user execute", are you suggesting to just do nothing, or something else?

Reply:
> Your message is longer than typical exception messages. Should that go to the logger, or into the exception?

You've actually answered this....

Reply:
OK I opted for a slightly shorter message (which I prefer), but took your suggestion. what do you think/

Reply:
?

Reply:
Yes that works for me. Please fixup in the source as well. Thx!

Reply:
> Please fixup in the source as well.

What do you mean by that?